chart/scripts/pre-upgrade-hook.sh

```diff
@@ -0,0 +1,131 @@
+#!/bin/bash
+
+set -eo pipefail
+
+# Global counters
+declare -A COUNTS
+RESOURCES_FOUND=false
+
+check_prerequisites() {
+  if ! command -v kubectl &>/dev/null; then
+    echo "Missing required tool: kubectl"
+    exit 1
+  fi
+}
+
+print_resource_table() {
+  local kind="$1"
+  local items="$2"
+  local -a headers=("${@:3}")
+
+  local count
+  count=$(wc -l <<< "$items")
+  COUNTS["$kind"]=$count
+  RESOURCES_FOUND=true
+
+  echo "Found $count $kind resource(s):"
+  echo
+
+  IFS=$'\n' read -r -d '' -a lines < <(printf '%s\0' "$items")
+
+  # Initialize max_lengths array with header lengths
+  local -a max_lengths
+  for i in "${!headers[@]}"; do
+    max_lengths[i]=${#headers[i]}
+  done
+
+  # Calculate max width for each column
+  for line in "${lines[@]}"; do
+    IFS=$'\t' read -r -a cols <<< "$line"
+    for i in "${!cols[@]}"; do
+      (( ${#cols[i]} > max_lengths[i] )) && max_lengths[i]=${#cols[i]}
+    done
+  done
+
+  for i in "${!headers[@]}"; do
+    printf "%-${max_lengths[i]}s  " "${headers[i]}"
+  done
+  printf "\n"
+
+  for i in "${!headers[@]}"; do
+    printf "%-${max_lengths[i]}s  " "$(printf '%*s' "${max_lengths[i]}" '' | tr ' ' '-')"
+  done
+  printf "\n"
+
+  for line in "${lines[@]}"; do
+    IFS=$'\t' read -r -a cols <<< "$line"
+    for i in "${!cols[@]}"; do
+      printf "%-${max_lengths[i]}s  " "${cols[i]}"
+    done
+    printf "\n"
+  done
+
+  echo
+}
+
+detect_resource() {
+  local crd="$1"
+  local kind="$2"
+  local jsonpath="$3"
+  local -a headers=("${@:4}")
+
+  echo "Checking for $kind resources..."
+
+  local output
+  if ! output=$(kubectl get "$crd" --all-namespaces -o=jsonpath="$jsonpath" 2>&1); then
+    if grep -q "the server doesn't have a resource type" <<< "$output"; then
+      echo "Resource type $crd not found. Skipping."
+      echo
+      return 0
+    else
+      echo "Error retrieving $kind resources: $output"
```

Comment:
We should consider adding retries or maybe bump `backoffLimit` for the job in case there's transient errors.